### ISSUE TITLE
Prediction result plot colors

### ISSUE THREAD

**Issue by altaris** [ultralytics/ultralytics#22732]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

When [plotting a prediction result](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Results.plot), the colors of the original image does not match if the input image was fed to the model as a PIL.Image object. However, if the input image was fed as a tensor, then the plot colors are correct.

In the example below, I run prediction on a completely red image. In `TEST 1`, a feed the image as a tensor, and [`Results.plot`](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Results.plot) correctly displays a red image. However, in `TEST 2`, i feed the image as a PIL.Image object, and [`Results.plot`](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Results.plot) shows a completely blue image.

### Environment

```
Ultralytics 8.3.204 🚀 Python-3.11.13 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 4090, 24072MiB)
Setup complete ✅ (32 CPUs, 124.9 GB RAM, 701.0/3665.4 GB disk)

OS                     Linux-6.14.0-27-generic-x86_64-with-glibc2.39
Environment            Linux
Python                 3.11.13
Install                git
Path                   <REDACTED>
RAM                    124.95 GB
Disk                   701.0/3665.4 GB
CPU                    AMD Ryzen 9 7950X3D 16-Core Processor
CPU count              32
GPU                    NVIDIA GeForce RTX 4090, 24072MiB
GPU count              2
CUDA                   12.8

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.10.1>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.1.0>=7.1.2
pyyaml                 ✅ 6.0.2>=5.3.1
requests               ✅ 2.32.3>=2.23.0
scipy                  ✅ 1.15.2>=1.4.1
torch                  ✅ 2.8.0>=1.8.0
torch                  ✅ 2.8.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0>=0.9.0
psutil                 ✅ 7.0.0
polars                 ✅ 1.34.0
ultralytics-thop       ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example

```py
from IPython.display import display
import torchvision.transforms.functional as tr
from ultralytics import YOLO
import torch
from PIL import Image

model = YOLO("yolo11n.pt")
image_array = torch.stack(
    [
        torch.ones((256, 256)),
        torch.zeros((256, 256)),
        torch.zeros((256, 256)),
    ]
)
image = tr.to_pil_image(image_array)
display(image)  # all red

# TEST 1 ========================================
r = model(image_array.unsqueeze(0), verbose=False)[0]
display(tr.to_pil_image(r.plot()))  # all red
display(Image.fromarray(r.plot()))  # all red
display(r.plot(pil=True))  # all red

# TEST 2 ========================================
r = model(image, verbose=False)[0]
display(tr.to_pil_image(r.plot()))  # all blue
display(Image.fromarray(r.plot()))  # all blue
display(r.plot(pil=True))  # all blue
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @altaris, thank you for your detailed report and for using Ultralytics 🚀! This is an automated response to help get things started; an Ultralytics engineer will review your issue and follow up with you soon 🙂

In the meantime, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples that may be relevant to prediction and plotting behavior.

Since this appears to be a 🐛 Bug Report related to `Results.plot` color handling, please ensure your example is a true [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE). Your snippet is already very helpful 🙌 — if you can further confirm that:
- it runs as-is from a clean environment, and  
- it reproduces the issue with no extra dependencies or notebook-only state  

that will make it even easier for engineers to debug.

If this evolves into a custom training or usage ❓ Question (for example, around preprocessing, color spaces, or transforms), please provide as much information as possible, including:
- exact code used for prediction and plotting  
- any preprocessing steps (transforms, conversions, etc.)  
- screenshots of the mismatched colors  

Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) if training is involved anywhere in your workflow.

Join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧  
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/)  
- To share knowledge with the community, explore our [Subreddit](https://www.reddit.com/r/Ultralytics/)  

## Upgrade

Before engineers triage this further, please upgrade to the latest `ultralytics` package (and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/), and confirm the behavior still occurs. This helps ensure the issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

If you can reproduce the PIL vs tensor color discrepancy in any of the environments above and note which one you used, that information will be very helpful for the engineering team.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
There's a bug. But it's the opposite. `r.plot()` returns BGR, so it's supposed to show red as blue if you don't reverse the channels.

The bug is with `r.plot(pil=True)`. It should return RGB, but currently it returns BGR.

And another bug is with `torch.Tensor` input which is not converted to BGR correctly.

The correct test code is:

```python
# TEST 1 ========================================
r = model(image_array.unsqueeze(0), verbose=False)[0]
display(tr.to_pil_image(r.plot()[..., ::-1]))
display(Image.fromarray(r.plot()[..., ::-1]))
display(r.plot(pil=True))

# TEST 2 ========================================
r = model(image, verbose=False)[0]
display(tr.to_pil_image(r.plot()[..., ::-1]))
display(Image.fromarray(r.plot()[..., ::-1]))
display(r.plot(pil=True))
```

**Comment by altaris** [comment]
Ok, so to confirm:

1. `Results.plot` returns a BGR image array,
2. `Results.plot` with `pil=True` returns a RGB image,
3. `model.predict` should receive a RGB image or image array.

**Comment by glenn-jocher** [comment]
Yes, that’s the intended contract: `Results.plot()` returns a BGR `ndarray` (as documented in the [Results.plot docs](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Results.plot)), `Results.plot(pil=True)` is meant to return an RGB PIL image (the BGR output there is the bug we’ll fix), and `model()`/`model.predict()` expect RGB inputs, so your summary is correct aside from that known bug in point 2.

**Comment by UltralyticsAssistant** [comment]
A potential fix for the color-conversion and plotting behavior discussed in this issue has now been merged in [PR #22734: `ultralytics 8.3.230` – Fix color conversion for Tensor inputs and PIL plots](https://github.com/ultralytics/ultralytics/pull/22734) by @Y-T-G with contributions from @glenn-jocher.  

### What changed and why it might fix this

The merged PR includes several updates that are directly related to color handling and plotting:

- `Results.plot()` has been standardized to always use `annotator.result(pil)`, so:
  - `Results.plot(pil=False)` now consistently returns a NumPy array.
  - `Results.plot(pil=True)` now consistently returns an RGB PIL image.
- `Annotator` internals were adjusted:
  - Internal images stay in BGR when working with PIL, matching the existing color palette.
  - `Annotator.result(pil=False)` returns a BGR NumPy array by default, and an RGB PIL image when `pil=True`.
- Postprocessing color ordering was fixed when `orig_imgs` is a tensor for:
  - RT-DETR
  - SAM
  - YOLO classification
  - YOLO detection  
  In these paths, tensor images are now properly converted and channel-flipped to correct BGR/RGB ordering before overlays and plotting.

These changes should resolve issues where visualizations appeared with incorrect colors or where `Results.plot()` returned unexpected types or color formats, especially when using tensor inputs.

### How to test the fix

The PR bumps the version to `8.3.230`. To test the changes immediately:

1. **Install or upgrade to the new version (once available on PyPI):**
   ```bash
   pip install -U ultralytics>=8.3.230
   ```

2. **Or install directly from the latest `main` branch:**
   ```bash
   pip install -U git+https://github.com/ultralytics/ultralytics.git@main
   ```

After upgrading, re-run the same workflow that previously produced incorrect colors or inconsistent `Results.plot()` outputs and check whether:

- Plotted images now match the original image colors.
- `Results.plot(pil=False)` returns a NumPy array with correct colors.
- `Results.plot(pil=True)` returns an RGB PIL image with correct colors.
- Tensor-based inputs (for RT-DETR, SAM, classification, detection) now produce correctly colored overlays and saved results.

### Feedback

Feedback on whether the issue is resolved with `8.3.230` (or the latest `main`) would be very helpful.  

🙏 Thanks for reporting this and helping improve the library. If any color or plotting issues persist after updating, further details and examples are very welcome so the behavior can be refined.
